lib/connections/service.go

```diff
@@ -693,3 +698,45 @@ func IsAllowedNetwork(host string, allowed []string) bool {
 
 	return false
 }
+
+func dialParallel(deviceID protocol.DeviceID, dialTargets []dialTarget) (internalConn, bool) {
+	// Group targets into buckets by priority
+	dialTargetBuckets := make(map[int][]dialTarget, len(dialTargets))
+	for _, tgt := range dialTargets {
+		dialTargetBuckets[tgt.priority] = append(dialTargetBuckets[tgt.priority], tgt)
+	}
+
+	// Get all available priorities
+	priorities := make([]int, 0, len(dialTargetBuckets))
+	for prio := range dialTargetBuckets {
+		priorities = append(priorities, prio)
+	}
+
+	// Sort the priorities so that we dial form lowest (which means highest...)
+	sort.Ints(priorities)
+
+	for _, prio := range priorities {
+		tgts := dialTargetBuckets[prio]
+		res := make(chan internalConn, len(tgts))
+		wg := sync.NewWaitGroup()
+		for _, tgt := range tgts {
+			wg.Add(1)
+			go tgt.Dial(deviceID, wg, res)
+		}
+
+		go func(deviceID protocol.DeviceID, prio int) {
+			wg.Wait()
+			close(res)
+			l.Debugln("discarding", len(res), "connections while connecting to", deviceID, prio)
+			for conn := range res {
+				conn.Close()
+			}
+		}(deviceID, prio)
+
+		if conn, ok := <-res; ok {
```

Comment:
Isn't this a race with range above in case there is only once dialer for this prio (or all except the last one fail)? As the channel is buffered, the send in the dialer will return immediately, the wait group condition is fulfilled and the channel closed. While unlikely, this could happen before this if condition receives a valid connection. Or is there something preventing this that I miss?

Reply:
Yes this can happen. This should be handled differently.